core/trino-main/src/main/java/io/trino/execution/scheduler/ExponentialGrowthPartitionMemoryEstimator.java

```diff
@@ -27,32 +29,64 @@
+                Ordering.natural().max(defaultMemoryLimit, getEstimatedMemoryUsage()),
                 false);
     }
 
     @Override
     public MemoryRequirements getNextRetryMemoryRequirements(Session session, MemoryRequirements previousMemoryRequirements, DataSize peakMemoryUsage, ErrorCode errorCode)
     {
         DataSize previousMemory = previousMemoryRequirements.getRequiredMemory();
-        DataSize baseMemory = Ordering.natural().max(peakMemoryUsage, previousMemory);
-        if (shouldIncreaseMemory(errorCode)) {
+
+        // start with the maximum of previously used memory and actual usage
+        DataSize newMemory = Ordering.natural().max(peakMemoryUsage, previousMemory);
+        if (isOutOfMemoryError(errorCode)) {
+            // multiply if we hit an oom error
             double growthFactor = getFaultTolerantExecutionTaskMemoryGrowthFactor(session);
-            return new MemoryRequirements(DataSize.of((long) (baseMemory.toBytes() * growthFactor), DataSize.Unit.BYTE), false);
+            newMemory = DataSize.of((long) (newMemory.toBytes() * growthFactor), DataSize.Unit.BYTE);
         }
-        return new MemoryRequirements(baseMemory, false);
+
+        // if we are still below current estimate for new partition let's bump further
+        newMemory = Ordering.natural().max(newMemory, getEstimatedMemoryUsage());
+
+        return new MemoryRequirements(newMemory, false);
     }
 
-    private boolean shouldIncreaseMemory(ErrorCode errorCode)
+    private boolean isOutOfMemoryError(ErrorCode errorCode)
     {
         return EXCEEDED_LOCAL_MEMORY_LIMIT.toErrorCode().equals(errorCode) // too many tasks from single query on a node
                 || CLUSTER_OUT_OF_MEMORY.toErrorCode().equals(errorCode); // too many tasks in general on a node
     }
 
     @Override
-    public void registerPartitionFinished(Session session, MemoryRequirements previousMemoryRequirements, DataSize peakMemoryUsage, boolean success, Optional<ErrorCode> errorCode) {}
+    public synchronized void registerPartitionFinished(Session session, MemoryRequirements previousMemoryRequirements, DataSize peakMemoryUsage, boolean success, Optional<ErrorCode> errorCode)
+    {
+        if (success) {
+            memoryUsageDistribution.add(peakMemoryUsage.toBytes());
+        }
+        if (!success && errorCode.isPresent() && isOutOfMemoryError(errorCode.get())) {
+            double growthFactor = SystemSessionProperties.getFaultTolerantExecutionTaskMemoryGrowthFactor(session);
+            // take previousRequiredBytes into account when registering failure on oom. It is conservative hence safer (and in-line with getNextRetryMemoryRequirements)
+            long previousRequiredBytes = previousMemoryRequirements.getRequiredMemory().toBytes();
+            long previousPeakBytes = peakMemoryUsage.toBytes();
+            memoryUsageDistribution.add(Math.max(previousRequiredBytes, previousPeakBytes) * growthFactor);
```

Comment:
Why would we add estimated memory usage to the distribution? If retry succeeds, then the actual usage will be added, right? This seems to skew the metric.

I understand your intention though. If one task consumes large amount of memory, then other tasks may also need large amount of memory. But this will make the stats collection inaccurate, maybe we should explore some other approach instead.

Reply:
Yeah - this surely is not exact science and I am not sure how well it will work in practice. But intention is exactly what you wrote. If we see that tasks are dying because we gave them too little memory we want to bump initial memory already for new tasks. Not wait until we have one which succeeds (it may take a lot of time till we have one).

I was thinking first about having two separate histograms for successful and unsuccessful tries. And make the one for unsuccessful decaying over time so "newer data is more important" - but I did not come up with reasonable way to merge the data from both, so I implemented the simple (yet I agree not 100% bullet-proof) approach.

Happy to hear suggestions how to improve though :)

BTW: I will add a commit on top with extra debug logging so we can see how it works in practice when testing out queries on cluster.

Reply:
Hmmm. Actually I guess with some tuning your approach might work well in practice. We can leave it as it is for now.